runtime/fresh/middlewares/3_main.ts

```diff
@@ -182,35 +183,41 @@ export const handler = [
 
     if (state?.flags.length > 0) {
       const currentCookies = getCookies(req.headers);
-      const segment = currentCookies[DECO_SEGMENT]
-        ? tryOrDefault(
-          () =>
-            JSON.parse(decodeURIComponent(atob(currentCookies[DECO_SEGMENT]))),
-          {},
-        )
-        : {};
+      const cookieSegment = tryOrDefault(
+        () => decodeCookie(currentCookies[DECO_SEGMENT]),
+        "",
+      );
+      const segment = tryOrDefault(() => JSON.parse(cookieSegment), {});
+
       const active = new Set(segment.active || []);
       const inactiveDrawn = new Set(segment.inactiveDrawn || []);
       for (const flag of state.flags) {
-        if (flag.value) {
-          active.add(flag.name);
-          inactiveDrawn.delete(flag.name);
-        } else {
-          active.delete(flag.name);
-          inactiveDrawn.add(flag.name);
+        // non-saved matcher includes ".rule" eg: "deco.matchers.1.rule"
+        // to avoid large cookies we only set saved matchers
+        const isSavedMatcher = !flag.name.includes(".rule");
```

Comment:
This is a very weak check, try to do this instead:

1) create a boolean value that starts with true and set to false at the very first loop right here => https://github.com/deco-cx/deco/blob/main/blocks/matcher.ts#L143

example

```tsx

let firstClass = true;
// loop
....
 // stop on first resolvable
if (type === "resolvable") {
  break;
}
firstClass = false;
}
```

Because in case of resolvable is reached in the first loop so the matcher is "saved" (lets say firstClass instead)

Reply:
Then, change the type flag to have a prop `firstClass?: boolean`

https://github.com/deco-cx/deco/blob/60d219ac0f5c6e5ede374c451a9f96818c41cc59/types.ts#L76

Reply:
then, push it right here: https://github.com/deco-cx/deco/blob/60d219ac0f5c6e5ede374c451a9f96818c41cc59/blocks/matcher.ts#L172

and reuse elsewhere